distributed/tests/test_core.py

```diff
@@ -605,30 +606,53 @@ async def ping(comm, delay=0.1):
     await rpc.close()
 
 
-@gen_test()
-async def test_connection_pool_close_while_connecting(monkeypatch):
-    """
-    Ensure a closed connection pool guarantees to have no connections left open
-    even if it is closed mid-connecting
-    """
-    from distributed.comm.registry import backends
-    from distributed.comm.tcp import TCPBackend, TCPConnector
+class WrongCancelConnector(TCPConnector):
+    async def connect(self, address, deserialize, **connection_args):
+        try:
+            await asyncio.sleep(10000)
+        except asyncio.CancelledError:
+            raise OSError("muhaha")
 
-    class SlowConnector(TCPConnector):
-        async def connect(self, address, deserialize, **connection_args):
+
+class WrongCancelBackend(TCPBackend):
+    _connector_class = WrongCancelConnector
+
+
+class SlowConnector(TCPConnector):
+    async def connect(self, address, deserialize, **connection_args):
+        try:
             await asyncio.sleep(10000)
-            return await super().connect(
-                address, deserialize=deserialize, **connection_args
-            )
+        except BaseException:
+            raise
 
-    class SlowBackend(TCPBackend):
-        _connector_class = SlowConnector
 
-    monkeypatch.setitem(backends, "tcp", SlowBackend())
+class SlowBackend(TCPBackend):
+    _connector_class = SlowConnector
+
+
+@pytest.mark.parametrize(
+    "backend",
+    [
+        SlowBackend,
+        WrongCancelBackend,
+    ],
+)
+@pytest.mark.parametrize(
+    "closing",
+    [
+        True,
+        False,
+    ],
+)
+@gen_test()
+async def test_connection_pool_cancellation(monkeypatch, closing, backend):
```

Comment:
These tests do not capture this race condition but the rewrite is much cleaner and covers more ground so I'd like to keep it